docs/writing-tests/making-a-testing-plan.md

```diff
@@ -0,0 +1,504 @@
+
+*Example* The following algorithm from [the DOM
+standard](https://dom.spec.whatwg.org/) powers
+[`document.querySelector`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector):
+
+> To **scope-match a selectors string** *selectors* against a *node*, run these
+> steps:
+>
+> 1. Let *s* be the result of [parse a
+>    selector](https://drafts.csswg.org/selectors-4/#parse-a-selector)
+>    *selectors*.
+>    [[SELECTORS4]](https://dom.spec.whatwg.org/#biblio-selectors4)
+> 2. If *s* is failure, then
+>    [throw](https://heycam.github.io/webidl/#dfn-throw) a
+>    "[`SyntaxError`](https://heycam.github.io/webidl/#syntaxerror)"
+>    [DOMException](https://heycam.github.io/webidl/#idl-DOMException).
+> 3. Return the result of [match a selector against a
+>    tree](https://drafts.csswg.org/selectors-4/#match-a-selector-against-a-tree)
+>    with *s* and *node*'s
+>    [root](https://dom.spec.whatwg.org/#concept-tree-root) using [scoping
+>    root](https://drafts.csswg.org/selectors-4/#scoping-root) *node*.
+>    [[SELECTORS4]](https://dom.spec.whatwg.org/#biblio-selectors4).
+
+As described earlier in this guide, we'd certainly want to test the branch
+regarding the parsing failure. However, there are many ways a string might fail
+to parse--should we verify them all in the tests for `document.querySelector`?
+What about `document.querySelectorAll`? Should we test them all there, too?
+
+The answers depend on the current state of the test suite: whether or not tests
+for selector parsing exist and where they are located. That's why it's best to
+confer with the people who are maintaining the tests.
+
+### Avoid excessive breadth
+
+Many algorithms operate on input that takes the form of a scalar value. When
+the set of values is finite, it can be tempting to test them all exhaustively.
+When the set is very large, test authors can reduce repetition by defining
+tests programmatically in loops.
+
+Using advanced control flow techniques to dynamically generate tests has a
+number of drawbacks. It makes the test suite more difficult to understand since
+readers have to mentally "unwind" the iteration to determine what is actually
+being verified. It also increases the risk of bugs in the tests. These bugs may
+not be obvious--they may not cause failures, and they may exercise fewer cases
+than intended.
+
+Generally speaking, such exhaustive approaches are unlikely to catch more bugs
+than a handful of carefully-chosen test cases. Although the risks of dynamic
+test generation may be tolerable in some specific cases, it's usually best to
+select the most interesting edge cases and move on.
+
+*Example* We can see this consideration in the very first step of the
+`Response` constructor from [the Fetch
+standard](https://fetch.spec.whatwg.org/)
+
+> The `Response`(*body*, *init*) constructor, when invoked, must run these
+> steps:
+>
+> 1. If *init*["`status`"] is not in the range `200` to `599`, inclusive, then
```

Comment:
I don't think this is a good example because it's not the sort of case where the temptation is big and the tradeoffs hard. In fact I've never reviewed a test where someone's tried to cover each number in a range. Much more commonly is a failure to test the boundary condition and one above/below those values.

I would instead look for an example of a test that unnecessary tries too many combinations of elements and attributes "just in case" without considering what combinations could be interesting. Tests with many subtests that pass in all browsers are a good place to go looking. When I see those in review, which happens, I'd like to be able to point to this section.

Reply:
To slightly contradict myself, I suppose it's a bad idea to use an example where there's a reasonable discussion to be had about the tradeoff, as that would distract from the point.

https://wpt.fyi/results/websockets/constructor/004.html is a bit much, but not excessive to the point that it's a problematic test.

Since I don't have an examples I think is more realistic and yet clear, I'll rescind my request :)